spec/stateful/job_control.py

```diff
@@ -266,6 +266,30 @@ def suspend_status(sh):
   expect_prompt(sh)
 
 
+@register(skip_shells=['zsh'])
+def no_spurious_tty_take(sh):
+  'A background job getting stopped by SIGTTIN should not disrupt foreground processes'
+  expect_prompt(sh)
+
+  sh.sendline('cat &')
+  time.sleep(0.1)  # seems necessary
+
+  # background cat should have been stopped by SIGTTIN immediately, but we don't
```

Comment:
`cat` should only be stopped if we gave it input?    

But we didn't do that yet in this test case?

---

Hmm but it does in bash too ... not sure I get it

Reply:
If I understand correctly, SIGTTIN will get sent as soon as a child calls `read(STDIN,...)` (assuming they don't have the tty), not just when it has bytes to return

Reply:
Confirmed this with strace:

```bash
$ strace cat &
[1] 390091
melvin@oil-dev:~/oil$ execve("/usr/bin/cat", ["cat"], 0x7ffe6bdc2580 /* 28 vars */) = 0
brk(NULL)                               = 0x55edbf10e000
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6ffc5d8000
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=22406, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 22406, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f6ffc5d2000
close(3)                                = 0
openat(AT_FDCWD, "/lib/x86_64-linux-gnu/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0Ps\2\0\0\0\0\0"..., 832) = 832
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
newfstatat(3, "", {st_mode=S_IFREG|0755, st_size=1922136, ...}, AT_EMPTY_PATH) = 0
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
mmap(NULL, 1970000, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f6ffc3f1000
mmap(0x7f6ffc417000, 1396736, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x26000) = 0x7f6ffc417000
mmap(0x7f6ffc56c000, 339968, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x17b000) = 0x7f6ffc56c000
mmap(0x7f6ffc5bf000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1ce000) = 0x7f6ffc5bf000
mmap(0x7f6ffc5c5000, 53072, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f6ffc5c5000
close(3)                                = 0
mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6ffc3ee000
arch_prctl(ARCH_SET_FS, 0x7f6ffc3ee740) = 0
set_tid_address(0x7f6ffc3eea10)         = 390094
set_robust_list(0x7f6ffc3eea20, 24)     = 0
rseq(0x7f6ffc3ef060, 0x20, 0, 0x53053053) = 0
mprotect(0x7f6ffc5bf000, 16384, PROT_READ) = 0
mprotect(0x55edbe3cc000, 4096, PROT_READ) = 0
mprotect(0x7f6ffc60a000, 8192, PROT_READ) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
munmap(0x7f6ffc5d2000, 22406)           = 0
getrandom("\x03\xdb\x9a\xad\xf7\xc6\xf6\xc4", 8, GRND_NONBLOCK) = 8
brk(NULL)                               = 0x55edbf10e000
brk(0x55edbf12f000)                     = 0x55edbf12f000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=3048928, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 3048928, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f6ffc000000
close(3)                                = 0
newfstatat(1, "", {st_mode=S_IFCHR|0620, st_rdev=makedev(0x88, 0x4), ...}, AT_EMPTY_PATH) = 0
newfstatat(0, "", {st_mode=S_IFCHR|0620, st_rdev=makedev(0x88, 0x4), ...}, AT_EMPTY_PATH) = 0
fadvise64(0, 0, 0, POSIX_FADV_SEQUENTIAL) = 0
mmap(NULL, 139264, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f6ffc3cc000
read(0, 0x7f6ffc3cd000, 131072)         = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGTTIN {si_signo=SIGTTIN, si_code=SI_KERNEL} ---
--- stopped by SIGTTIN ---
```